src/Covenant/Internal/ASGBuilder.hs

```diff
@@ -53,154 +49,193 @@ makeFieldLabelsNoPrefix ''ASGBuilderState
-      sixArg :: Int -> GenT (Reader Word64) (ASGBuilder Id)
-      sixArg size = do
-        r1 <- argOrId size
-        r2 <- argOrId size
-        r3 <- argOrId size
-        r4 <- argOrId size
-        r5 <- argOrId size
-        r6 <- argOrId size
-        f <- liftGen arbitrary
-        pure $
-          r1 >>= \r1' ->
-            r2 >>= \r2' ->
-              r3 >>= \r3' ->
-                r4 >>= \r4' ->
-                  r5 >>= \r5' -> r6 >>= \r6' -> prim (PrimCallSix f r1' r2' r3' r4' r5' r6')
-      aLam :: Int -> GenT (Reader Word64) (ASGBuilder Id)
-      aLam size = do
-        body <- expand (argOrId size)
-        -- Note (Koz, 21/01/25): We 'bypass' our protections on construction of
-        -- lambdas here for convenience, as otherwise, we would have to carry
-        -- around a `Scope n`, with `n` changing. While not impossible, this
-        -- would require existentialization (using something like `Some`), which
-        -- doesn't really help us any (because we have full control anyway), but
-        -- does make the code much harder to read and debug.
-        pure $ body >>= \bodyRef -> idOf (Lam bodyRef)
-      -- Note (Koz, 21/01/25): This is essentially `local`, but because GenT
-      -- isn't an instance of MonadReader, we have to write it by hand. We could
-      -- have defined an orphan instance instead, but I figured it'd be better
-      -- not to.
-      expand ::
-        forall (a :: Type).
-        GenT (Reader Word64) (ASGBuilder a) -> GenT (Reader Word64) (ASGBuilder a)
-      expand (GenT f) = GenT $ \qcgen size -> local (+ 1) (f qcgen size)
-      anApp :: Int -> GenT (Reader Word64) (ASGBuilder Id)
-      anApp size = do
-        lhs <- aLam size
-        rhs <- argOrId size
-        -- Note (Koz, 21/01/25): This is technically not as general as it could
-        -- be. By 'recycling' aLam, we always generate a 'fresh' lambda, rather
-        -- than (possibly) re-using one available to us as an argument. This is
-        -- not really practical to track at the moment, though once we have
-        -- types, it could be done. For now, this works well enough.
-        pure $ lhs >>= \lhsRef -> rhs >>= \rhsRef -> app (AnId lhsRef) rhsRef
-      argOrId :: Int -> GenT (Reader Word64) (ASGBuilder Ref)
-      argOrId originalSize = do
-        availableArgs <- lift ask
-        if availableArgs == 0
-          -- No arguments possible, so reduce the size and recurse
-          then fmap AnId <$> go (originalSize `quot` 2)
-          -- 20% chance we make an argument, 80% chance we don't
-          else
-            frequency
-              [ (1, pure . AnArg . Arg <$> liftGen (chooseEnum (0, availableArgs - 1))),
-                (4, fmap AnId <$> go (originalSize `quot` 2))
-              ]
+runASGBuilder :: ASGBuilder a -> ASGBuilderState -> (Either ASGBuilderError a, ASGBuilderState)
+runASGBuilder (ASGBuilder m) s =
+  let stateM = runExceptT m
+   in runState stateM s
```

Comment:
This is a bit different to the original, and unfortunately isn't doing quite what we want. The reason why the original builder runner worked the way it did was because, by _forcing_ the monad to produce a result of type `Id`, we ensure we have a toplevel computation, which this does not do.

Reply:
Do you mean we should restrict the type of `runASGBuilder` to accept `ASGBuilder Id` or we should flip the order of `State` and `Either` in the stack? 
If it's the latter case, it would just make the partial state unavailable in the case of an error, right? It wouldn't ensure we have a top-level computation?